cpp/src/DataStorm/SessionI.cpp

```diff
@@ -1134,28 +1134,37 @@ SessionI::subscriberInitialized(
         out << _id << ": initialized '" << element << "' from 'e" << elementId << '@' << topicId << "'";
     }
     elementSubscriber->initialized = true;
-    elementSubscriber->lastId = samples.empty() ? 0 : samples.back().id;
 
-    vector<shared_ptr<Sample>> samplesI;
-    samplesI.reserve(samples.size());
-    auto sampleFactory = element->getTopic()->getSampleFactory();
-    auto keyFactory = element->getTopic()->getKeyFactory();
-    for (const auto& sample : samples)
+    if (samples.empty())
     {
-        assert((!key && !sample.keyValue.empty()) || key == subscriber.keys[sample.keyId].first);
-
-        samplesI.push_back(sampleFactory->create(
-            _id,
-            elementSubscribers->name,
-            sample.id,
-            sample.event,
-            key ? key : keyFactory->decode(_instance->getCommunicator(), sample.keyValue),
-            subscriber.tags[sample.tag],
-            sample.value,
-            sample.timestamp));
-        assert(samplesI.back()->key);
+        return {};
```

Comment:
In the original code, when samples.empty(), we set elementSubscriber->lastId to 0. 

It's not immediately clear why we don't need that. Is this lastId already 0 for some other reason?

Reply:
I added a comment to explain the logic of `lastId`.

lastId is default initialized to 0 in Session.h

```
// The ID of the last processed sample.
std::int64_t lastId{0};
```

Reply:
Ok, and the fix in this PR is exactly that: to not set lastId to 0 when samples is empty?

Reply:
The fix is not reset to 0 when the subscriber is initialized after recovery. 

The subscriber received some samples and lastId is updated accordingly.

Then the Session is lost, when it reconnects subscriberInitialized is called again.

If the new call sent no samples, because there were no new samples since the recovery, the previous code was reseting lastId to 0. (that is the bug).

Now if session is lost again, the next recovery would tell the peer that the lastId it saw is 0, and the peer would send all queues elements. That is what was happening with the test failure.

Reply:
Pushed an additional test that allows reproducing the initial issue.